08.ls_object/lib/ls_long.rb

```diff
@@ -0,0 +1,106 @@
+# frozen_string_literal: true
+
+class LsLong
+  attr_reader :file
+
+  def initialize(files)
+    @files = files
+  end
+
+  def call
+    files_detail = @files.each.map do |file|
+      build_file_data(file)
+    end
+    max_lengths = find_max_text(files_detail)
+    total = files_detail.map do |file|
+      file[:block].to_i
+    end.sum
+    total_line = "total #{total}"
+    body = files_detail.map do |file|
+      formatted_for_display(file, *max_lengths)
+    end
+    [total_line, *body].join("\n")
+  end
+
+  private
+
+  def build_file_data(file)
+    {
+      type: filetypes(file),
+      permission: permissions(file),
+      links: File.lstat(file).nlink.to_s,
+      user: Etc.getpwuid(File.lstat(file).uid).name,
+      group: Etc.getgrgid(File.lstat(file).gid).name,
+      size: File.lstat(file).size.to_s,
+      timestamp: timestamp(file),
+      file: File.basename(file),
+      linked_file: linked_file(file),
+      block: File.lstat(file).blocks.to_i
+    }
+  end
+
+  def filetypes(file)
+    case File.lstat(file).ftype
+    when 'link'
+      'l'
+    when 'directory'
+      'd'
+    else
+      '-'
+    end
+  end
+
+  def permissions(file)
+    numbers = File.lstat(file).mode.to_s(8)[-3..-1].chars
+    permissions = []
+    numbers.each do |number|
+      permissions << permission(:"#{number}")
+    end
+    permissions.join
+  end
+
+  def permission(num)
+    {
+      '7': 'rwx',
+      '6': 'rw-',
+      '5': 'r-x',
+      '4': 'r--',
+      '3': '-wx',
+      '2': '-w-',
+      '1': '--x',
+      '0': '---'
+    }[num]
+  end
+
+  def timestamp(file)
+    timestamps = File.lstat(file).mtime
+    "#{timestamps.strftime('%-m').rjust(2)} #{timestamps.strftime('%e %H:%M')}"
+  end
+
+  def linked_file(file)
+    " -> #{File.readlink(file)}" if filetypes(file) == 'l'
+  end
+
+  def find_max_text(files)
+    keys = %i[links user group size]
+    keys.map do |key|
+      files.map do |file|
+        file[key].size
+      end.max
+    end
+  end
+
+  def formatted_for_display(file, links_max_length, user_max_length, group_max_length, size_max_length)
+    [
+      file[:type],
+      file[:permission],
+      "  #{file[:links].rjust(links_max_length)}",
+      " #{file[:user].ljust(user_max_length)}",
+      "  #{file[:group].ljust(group_max_length)}",
+      "  #{file[:size].rjust(size_max_length)}",
+      " #{file[:timestamp]}",
+      " #{file[:file]}",
+      (file[:linked_file]).to_s
```

Comment:
```suggestion
      file[:linked_file]
```

`#linked_file(file)`の戻り値を必ずStringにしておけばよさそうです。

Reply:
どの場合でもStringを返すようにしたので削除しました！`linked_file`メソッドについては次のコメントで修正内容記載しています！